File: lint-arwaky/archive/taxonomy/app_config_vo.py
"""app_config_vo — Unified configuration for the application."""

from __future__ import annotations
import os
from typing import TYPE_CHECKING
from pydantic import BaseModel, ConfigDict

from .log_suggestion_vo import BooleanVO
from .file_path_vo import DirectoryPath
from .adapter_name_vo import AdapterName
from .adapter_collection_vo import AdapterNameList

if TYPE_CHECKING:
    from .config_setting_vo import ProjectConfig, Thresholds, AdapterStatus


class AppConfig(BaseModel):
    """Unified configuration — transport, paths, and project settings."""

    model_config = ConfigDict(arbitrary_types_allowed=True, frozen=True)

    phantom_root: DirectoryPath
    project_root: DirectoryPath
    project: "ProjectConfig"

    @classmethod
    def create(
        cls,
        phantom_root: str | None = None,
        project_root: str | None = None,
        project: "ProjectConfig" | None = None,
    ) -> AppConfig:
        from .config_setting_vo import ProjectConfig

        p_root = phantom_root or os.environ.get("PHANTOM_ROOT", os.path.expanduser("~"))
        proj_root = project_root or os.environ.get("PROJECT_ROOT", os.getcwd())
        proj = project or ProjectConfig.defaults()

        return cls(
            phantom_root=DirectoryPath(value=p_root),
            project_root=DirectoryPath(value=proj_root),
            project=proj,
        )

    @property
    def thresholds(self) -> "Thresholds":
        return self.project.thresholds

    def adapter_status(self, name: str) -> "AdapterStatus":
        """Get status for a named adapter."""
        from .config_setting_vo import AdapterStatus

        for entry in self.project.adapters:
            if entry.name == name:
                return entry.status
        return AdapterStatus.NOT_INSTALLED

    def is_adapter_enabled(self, name: str | AdapterName) -> BooleanVO:
        from .config_setting_vo import AdapterStatus

        status = self.adapter_status(str(name))
        return BooleanVO(value=status == AdapterStatus.ENABLED)

    def active_adapters(self) -> AdapterNameList:
        """Names of enabled adapters."""
        return AdapterNameList(
            values=[
                AdapterName(value=e.name) for e in self.project.adapters if e.is_active
            ]
        )

    def __repr__(self) -> str:
        return (
            f"AppConfig("
            f"phantom={self.phantom_root!r}, "
            f"adapters={self.active_adapters()})"
        )

File: lint-arwaky/src-rust/taxonomy/app_config_vo.rs
use std::env;

use crate::taxonomy::{
    AdapterName, AdapterNameList, BooleanVO, DirectoryPath,
};

/// app_config_vo — Unified configuration for the application.

/// Unified configuration — transport, paths, and project settings.
#[derive(Debug, Clone)]
pub struct AppConfig {
    phantom_root: DirectoryPath,
    project_root: DirectoryPath,
    project: crate::taxonomy::ProjectConfig,
}

impl AppConfig {
    /// Create a new AppConfig.
    ///
    /// # Arguments
    /// * `phantom_root` - Optional phantom root directory. Defaults to environment variable `PHANTOM_ROOT` or home directory.
    /// * `project_root` - Optional project root directory. Defaults to environment variable `PROJECT_ROOT` or current directory.
    /// * `project` - Optional project configuration. Defaults to `ProjectConfig::defaults()`.
    pub fn create(
        phantom_root: Option<&str>,
        project_root: Option<&str>,
        project: Option<crate::taxonomy::ProjectConfig>,
    ) -> Self {
        let p_root = phantom_root
            .or_else(|| env::var("PHANTOM_ROOT").ok())
            .unwrap_or_else(|| env::var("HOME").unwrap_or_else(|_| ".".to_string()));
        let proj_root = project_root
            .or_else(|| env::var("PROJECT_ROOT").ok())
            .unwrap_or_else(|| env::current_dir().unwrap().to_string_lossy().to_string());
        let proj = project.unwrap_or_else(crate::taxonomy::ProjectConfig::defaults);

        Self {
            phantom_root: DirectoryPath::new(p_root).expect("Invalid phantom root"),
            project_root: DirectoryPath::new(proj_root).expect("Invalid project root"),
            project: proj,
        }
    }

    /// Get the thresholds from the project configuration.
    pub fn thresholds(&self) -> &crate::taxonomy::Thresholds {
        &self.project.thresholds
    }

    /// Get status for a named adapter.
    pub fn adapter_status(&self, name: &str) -> crate::taxonomy::AdapterStatus {
        for entry in &self.project.adapters {
            if entry.name == name {
                return entry.status;
            }
        }
        crate::taxonomy::AdapterStatus::NOT_INSTALLED
    }

    /// Check if an adapter is enabled.
    pub fn is_adapter_enabled(&self, name: &str) -> BooleanVO {
        let status = self.adapter_status(name);
        BooleanVO::new(status == crate::taxonomy::AdapterStatus::ENABLED)
            .expect("Invalid boolean value")
    }

    /// Names of enabled adapters.
    pub fn active_adapters(&self) -> AdapterNameList {
        let mut values = Vec::new();
        for entry in &self.project.adapters {
            if entry.is_active {
                values.push(AdapterName::new(entry.name.clone())
                    .expect("Invalid adapter name"));
            }
        }
        AdapterNameList { values }
    }
}

impl std::fmt::Display for AppConfig {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "AppConfig(phantom={}, adapters={:?})",
            self.phantom_root,
            self.active_adapters()
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::taxonomy::{AdapterStatus, ProjectConfig, Thresholds};

    #[test]
    fn test_app_config_create() {
        let config = AppConfig::create(
            Some("/phantom"),
            Some("/project"),
            Some(ProjectConfig::defaults()),
        );
        assert_eq!(config.phantom_root.to_string(), "/phantom");
        assert_eq!(config.project_root.to_string(), "/project");
    }

    #[test]
    fn test_app_config_defaults() {
        // Set environment variables for deterministic test
        env::set_var("PHANTOM_ROOT", "/test/phantom");
        env::set_var("PROJECT_ROOT", "/test/project");
        let config = AppConfig::create(None, None, None);
        assert_eq!(config.phantom_root.to_string(), "/test/phantom");
        assert_eq!(config.project_root.to_string(), "/test/project");
        // Clean up
        env::remove_var("PHANTOM_ROOT");
        env::remove_var("PROJECT_ROOT");
    }
}
